mqtt/mqtt-broker/src/persist.rs

```diff
@@ -1,28 +1,385 @@
+/// copies of the state and rollback in case something gets corrupted.
+/// It also aids in "transactionally" saving the state. Updating the symlink
+/// "commits" the changes. This is an attempt to prevent corrupting the state
+/// file in case the process crashes in the middle of writing.
+#[derive(Debug)]
+pub struct FilePersistor<F> {
+    dir: PathBuf,
+    format: F,
+    previous_count: usize,
+
+    /// seq is bumped on every store to disambiguate stores in the same timestamp
+    seq: u16,
+}
+
+impl<F> FilePersistor<F> {
+    pub fn new<P: Into<PathBuf>>(dir: P, format: F) -> Self {
+        FilePersistor {
+            dir: dir.into(),
+            format,
+            previous_count: STATE_DEFAULT_PREVIOUS_COUNT,
+            seq: 0,
+        }
+    }
+
+    /// Sets the number of past states to save
+    ///
+    /// The intent is to allow rollback to a pervious state.
+    /// The default is `2`, which saves the current and previous state.
+    /// The minimum value is `1` (the current state).
+    pub fn with_previous_count(mut self, previous_count: usize) -> Self {
+        self.previous_count = cmp::max(1, previous_count);
+        self
+    }
+}
+
+/// A simple format based on Bincode and serde
+#[derive(Clone, Debug, Default)]
+pub struct BincodeFormat;
+
+impl BincodeFormat {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+impl FileFormat for BincodeFormat {
+    type Error = Error;
+
+    fn load<R: Read>(&self, reader: R) -> Result<BrokerState, Self::Error> {
+        let decoder = GzDecoder::new(reader);
+        fail_point!("bincodeformat.load.deserialize_from", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        let state = bincode::deserialize_from(decoder)
+            .context(ErrorKind::Persist(ErrorReason::Deserialize))?;
+        Ok(state)
+    }
+
+    fn store<W: Write>(&self, writer: W, state: BrokerState) -> Result<(), Self::Error> {
+        let encoder = GzEncoder::new(writer, Compression::default());
+        fail_point!("bincodeformat.store.serialize_into", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        bincode::serialize_into(encoder, &state)
+            .context(ErrorKind::Persist(ErrorReason::Serialize))?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<F> Persist for FilePersistor<F>
+where
+    F: FileFormat + Clone + Send + 'static,
+{
+    type Error = Error;
+
+    async fn load(&mut self) -> Result<Option<BrokerState>, Self::Error> {
+        let dir = self.dir.clone();
+        let format = self.format.clone();
+
+        let res = tokio::task::spawn_blocking(move || {
+            let path = dir.join(format!("{}.{}", STATE_DEFAULT_STEM, STATE_EXTENSION));
+            if path.exists() {
+                info!("loading state from file {}.", path.display());
+
+                fail_point!("filepersistor.load.fileopen", |_| {
```

Comment:
It is a seam for us to emulate failure during following operation in tests, right?

Reply:
Yes, this is a test only thing. The macro doesn't generate any code in normal operation. The test code turns on a feature that causes code to be generated to enable fault injection.